File: ToyORBC#/ToyORB/ToyORB.cs
using System.Dynamic;
using System.Net;
using System.Net.Sockets;
using System.Reflection;
using System.Text;
using System.Threading.Channels;

namespace ToyORB;

public class ToyOrb
{
    public static readonly Address ToyOrbAddress = new Address("127.0.0.1", 2000);

    private static ToyOrb? _instance;
    private ToyOrb() {
        // Create a new instance of the ToyORB
    }

    public static ToyOrb? GetInstance() {
        if(_instance==null) {
            _instance = new ToyOrb();
        }
        return _instance;
    }
    public int register(string name, object obj) {
        // Register the object with the name in the registry
        
        Message msg=new Message(name,obj.GetType().GetInterfaces()[0].Name);
        //byte[] resp=r.deliver_and_wait_feedback(toyORBAddress, Marshaller.marshal(msg));
        TcpClient client = new TcpClient(ToyOrbAddress.dest(), ToyOrbAddress.port());
        NetworkStream stream = client.GetStream();
        var messageBytes = Marshaller.marshal(msg);
        stream.Write(messageBytes, 0, messageBytes.Length);
        Console.WriteLine($"Socket client sent message:\"{msg.data+" "+msg.sender}\"");
        // Receive ack.
        var buffer = new byte[1024];
        var received = stream.Read(buffer, 0, buffer.Length);
        var response = Marshaller.unmarshal(buffer);
        Console.WriteLine("ToyORB registered server at "+response.data);
        return Int32.Parse(response.data.Split(":")[1].Trim());
    }

    public object GetObjectRef(String name)
    {
        Message msg = new Message(name, "");
        TcpClient client = new TcpClient(ToyOrbAddress.dest(), ToyOrbAddress.port());
        NetworkStream stream = client.GetStream();
        var messageBytes = Marshaller.marshal(msg);
        stream.Write(messageBytes, 0, messageBytes.Length);
        Console.WriteLine($"Socket client sent message:\"{msg.data+" "+msg.sender}\"");

        // Receive ack.
        var buffer = new byte[1024];
        var received = stream.Read(buffer, 0, buffer.Length);
        var response = Marshaller.unmarshal(buffer);
        // object proxy = Create<ICalculator, CalculatorProxy>();
        // ((CalculatorProxy)proxy)._calculator = calculator;
        // return (ICalculator)proxy;
        string interfaceName=response.data.Split(",")[1];
        int port =Int32.Parse(response.data.Split(",")[0].Split(':')[1].Trim());
        Console.WriteLine($"interfaceName: {interfaceName}, port: {port}");
        return ServiceProxy.CreateProxy(interfaceName, port);
    }
    
}

File: toyORBJava/src/ToyORB.java
import MessageMarshaller.Marshaller;
import MessageMarshaller.Message;
import Registry.Entry;
import RequestReply.Requestor;

import java.lang.constant.DynamicCallSiteDesc;
import java.lang.reflect.Proxy;

public class ToyORB {
    public final Entry toyORBAddress=new Entry("127.0.0.1", 2000);

    private static ToyORB instance;
    private ToyORB() {
        // Create a new instance of the ToyORB
    }

    public static ToyORB getInstance() {
        if(instance==null) {
            instance = new ToyORB();
        }
        return instance;
    }

    private Requestor r=new Requestor("ToyORB");
    public int register(String name, Object obj) {
        // Register the object with the name in the registry
        Message msg=new Message(name,obj.getClass().getInterfaces()[0].getName());
        byte[] resp=r.deliver_and_wait_feedback(toyORBAddress, Marshaller.marshal(msg));
        System.out.println("ToyORB registered server at "+Marshaller.unmarshal(resp).data);
        return Integer.parseInt(Marshaller.unmarshal(resp).data.split(":")[1].trim());
    }

    public Object getObjectRef(String name){
        Message msg=new Message(name,"");
        byte[] resp=r.deliver_and_wait_feedback(toyORBAddress, Marshaller.marshal(msg));
        if(Marshaller.unmarshal(resp).data.equals("null")){
            System.out.println("ToyORB could not find server");
            return null;
        }
        System.out.println("ToyORB found server at "+Marshaller.unmarshal(resp).data);
        String interfaceName=Marshaller.unmarshal(resp).data.split(",")[1];
        Class proxyclass;
        try{
            proxyclass=Class.forName(interfaceName);
        }catch(Exception e) {
            System.out.println("Error creating proxy class");
            return null;
        }
        return Proxy.newProxyInstance(
                DynamicProxy.class.getClassLoader(),
                new Class[]{proxyclass},
                new DynamicProxy(Integer.parseInt(Marshaller.unmarshal(resp).data.split(",")[0].split(":")[1].trim()))
        );
    }
}
